### INITIAL DATA CLEANING

Importing the original data file to investigate and clean the columns and row values.

In [47]:
#Importing the required packages
import pandas as pd
import matplotlib.pyplot as plt
import scipy.stats as st
import numpy as np

In [48]:
#Save the dataset file path
housing_path = "Property Sales of Melbourne City.csv"

In [49]:
#Read the csv file into a dataframe
housing_df = pd.read_csv(housing_path)

In [50]:
#Display the data table for preview
housing_df.head()

Unnamed: 0      Suburb           Address  Rooms Type    Price Method  \
0           1  Abbotsford      85 Turner St      2    h  1480000      S   
1           2  Abbotsford   25 Bloomburg St      2    h  1035000      S   
2           4  Abbotsford      5 Charles St      3    h  1465000     SP   
3           5  Abbotsford  40 Federation La      3    h   850000     PI   
4           6  Abbotsford       55a Park St      4    h  1600000     VB   

  SellerG       Date  Distance  ...  Bathroom  Car  Landsize  BuildingArea  \
0  Biggin  3/12/2016       2.5  ...       1.0  1.0     202.0           NaN   
1  Biggin  4/02/2016       2.5  ...       1.0  0.0     156.0          79.0   
2  Biggin  4/03/2017       2.5  ...       2.0  0.0     134.0         150.0   
3  Biggin  4/03/2017       2.5  ...       2.0  1.0      94.0           NaN   
4  Nelson  4/06/2016       2.5  ...       1.0  2.0     120.0         142.0   

   YearBuilt  CouncilArea  Lattitude Longtitude             Regionname  \
0        NaN        Yarra   -37.7996   144.9984  Northern Metropolitan   
1     1900.0        Yarra   -37.8079   144.9934  Northern Metropolitan   
2     1900.0        Yarra   -37.8093   144.9944  Northern Metropolitan   
3        NaN        Yarra   -37.7969   144.9969  Northern Metropolitan   
4     2014.0        Yarra   -37.8072   144.9941  Northern Metropolitan   

   Propertycount  
0         4019.0  
1         4019.0  
2         4019.0  
3         4019.0  
4         4019.0  

[5 rows x 22 columns]

In [51]:
#Provide a concise summary of the dataframe
housing_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18396 entries, 0 to 18395
Data columns (total 22 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Unnamed: 0     18396 non-null  int64  
 1   Suburb         18396 non-null  object 
 2   Address        18396 non-null  object 
 3   Rooms          18396 non-null  int64  
 4   Type           18396 non-null  object 
 5   Price          18396 non-null  int64  
 6   Method         18396 non-null  object 
 7   SellerG        18396 non-null  object 
 8   Date           18396 non-null  object 
 9   Distance       18395 non-null  float64
 10  Postcode       18395 non-null  float64
 11  Bedroom2       14927 non-null  float64
 12  Bathroom       14925 non-null  float64
 13  Car            14820 non-null  float64
 14  Landsize       13603 non-null  float64
 15  BuildingArea   7762 non-null   float64
 16  YearBuilt      8958 non-null   float64
 17  CouncilArea    12233 non-null  object 
 18  Lattit

In [52]:
#Remove the old indexing column 'Unnamed: 0'
del housing_df["Unnamed: 0"]

#Check the changes have applied
housing_df.head()

Suburb           Address  Rooms Type    Price Method SellerG  \
0  Abbotsford      85 Turner St      2    h  1480000      S  Biggin   
1  Abbotsford   25 Bloomburg St      2    h  1035000      S  Biggin   
2  Abbotsford      5 Charles St      3    h  1465000     SP  Biggin   
3  Abbotsford  40 Federation La      3    h   850000     PI  Biggin   
4  Abbotsford       55a Park St      4    h  1600000     VB  Nelson   

        Date  Distance  Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/12/2016       2.5    3067.0  ...       1.0  1.0     202.0           NaN   
1  4/02/2016       2.5    3067.0  ...       1.0  0.0     156.0          79.0   
2  4/03/2017       2.5    3067.0  ...       2.0  0.0     134.0         150.0   
3  4/03/2017       2.5    3067.0  ...       2.0  1.0      94.0           NaN   
4  4/06/2016       2.5    3067.0  ...       1.0  2.0     120.0         142.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
1     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
2     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   
3        NaN        Yarra  -37.7969    144.9969  Northern Metropolitan   
4     2014.0        Yarra  -37.8072    144.9941  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [53]:
#Change the 'Postcode' column to object data type values.
housing_df['Postcode'] = housing_df['Postcode'].astype('object')

#Check the changes have applied
housing_df.head()

Suburb           Address  Rooms Type    Price Method SellerG  \
0  Abbotsford      85 Turner St      2    h  1480000      S  Biggin   
1  Abbotsford   25 Bloomburg St      2    h  1035000      S  Biggin   
2  Abbotsford      5 Charles St      3    h  1465000     SP  Biggin   
3  Abbotsford  40 Federation La      3    h   850000     PI  Biggin   
4  Abbotsford       55a Park St      4    h  1600000     VB  Nelson   

        Date  Distance Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0  3/12/2016       2.5   3067.0  ...       1.0  1.0     202.0           NaN   
1  4/02/2016       2.5   3067.0  ...       1.0  0.0     156.0          79.0   
2  4/03/2017       2.5   3067.0  ...       2.0  0.0     134.0         150.0   
3  4/03/2017       2.5   3067.0  ...       2.0  1.0      94.0           NaN   
4  4/06/2016       2.5   3067.0  ...       1.0  2.0     120.0         142.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
1     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
2     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   
3        NaN        Yarra  -37.7969    144.9969  Northern Metropolitan   
4     2014.0        Yarra  -37.8072    144.9941  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [61]:
#Change the 'Date' column to datetime data type values.
housing_df['Date'] = housing_df['Date'].astype('datetime64[ns]')

#Check the changes have applied
housing_df.head()

Suburb           Address  Rooms Type    Price Method SellerG  \
0  Abbotsford      85 Turner St      2    h  1480000      S  Biggin   
1  Abbotsford   25 Bloomburg St      2    h  1035000      S  Biggin   
2  Abbotsford      5 Charles St      3    h  1465000     SP  Biggin   
3  Abbotsford  40 Federation La      3    h   850000     PI  Biggin   
4  Abbotsford       55a Park St      4    h  1600000     VB  Nelson   

        Date  Distance Postcode  ...  Bathroom  Car  Landsize  BuildingArea  \
0 2016-03-12       2.5   3067.0  ...       1.0  1.0     202.0           NaN   
1 2016-04-02       2.5   3067.0  ...       1.0  0.0     156.0          79.0   
2 2017-04-03       2.5   3067.0  ...       2.0  0.0     134.0         150.0   
3 2017-04-03       2.5   3067.0  ...       2.0  1.0      94.0           NaN   
4 2016-04-06       2.5   3067.0  ...       1.0  2.0     120.0         142.0   

   YearBuilt  CouncilArea Lattitude  Longtitude             Regionname  \
0        NaN        Yarra  -37.7996    144.9984  Northern Metropolitan   
1     1900.0        Yarra  -37.8079    144.9934  Northern Metropolitan   
2     1900.0        Yarra  -37.8093    144.9944  Northern Metropolitan   
3        NaN        Yarra  -37.7969    144.9969  Northern Metropolitan   
4     2014.0        Yarra  -37.8072    144.9941  Northern Metropolitan   

  Propertycount  
0        4019.0  
1        4019.0  
2        4019.0  
3        4019.0  
4        4019.0  

[5 rows x 21 columns]

In [55]:
#Summary Table of the numeric variables
housing_df.describe()

Rooms         Price      Distance      Bedroom2      Bathroom  \
count  18396.000000  1.839600e+04  18395.000000  14927.000000  14925.000000   
mean       2.935040  1.056697e+06     10.389986      2.913043      1.538492   
std        0.958202  6.419217e+05      6.009050      0.964641      0.689311   
min        1.000000  8.500000e+04      0.000000      0.000000      0.000000   
25%        2.000000  6.330000e+05      6.300000      2.000000      1.000000   
50%        3.000000  8.800000e+05      9.700000      3.000000      1.000000   
75%        3.000000  1.302000e+06     13.300000      3.000000      2.000000   
max       12.000000  9.000000e+06     48.100000     20.000000      8.000000   

                Car       Landsize  BuildingArea    YearBuilt     Lattitude  \
count  14820.000000   13603.000000   7762.000000  8958.000000  15064.000000   
mean       1.615520     558.116371    151.220219  1965.879996    -37.809849   
std        0.955916    3987.326586    519.188596    37.013261      0.081152   
min        0.000000       0.000000      0.000000  1196.000000    -38.182550   
25%        1.000000     176.500000     93.000000  1950.000000    -37.858100   
50%        2.000000     440.000000    126.000000  1970.000000    -37.803625   
75%        2.000000     651.000000    174.000000  2000.000000    -37.756270   
max       10.000000  433014.000000  44515.000000  2018.000000    -37.408530   

         Longtitude  Propertycount  
count  15064.000000   18395.000000  
mean     144.996338    7517.975265  
std        0.106375    4488.416599  
min      144.431810     249.000000  
25%      144.931193    4294.000000  
50%      145.000920    6567.000000  
75%      145.060000   10331.000000  
max      145.526350   21650.000000

In [56]:
#List of object type variables
object_columns = housing_df.select_dtypes("object")

#Summary Array of the object variables
object_columns.nunique()

Suburb           330
Address        18134
Type               3
Method             5
SellerG          305
Postcode         205
CouncilArea       33
Regionname         8
dtype: int64

In [57]:
#What is the date range in which these properties were sold? 
# 58 unique dates in the dataset.
housing_df['Date'].sort_values()

7900    2016-01-28
7899    2016-01-28
4741    2016-03-09
4368    2016-03-09
4369    2016-03-09
           ...    
14054   2017-12-08
14055   2017-12-08
14056   2017-12-08
14043   2017-12-08
14146   2017-12-08
Name: Date, Length: 18396, dtype: datetime64[ns]

In [58]:
#Looking into the duplication of street property addresses; 
# Are these due to the same streets occurring in different suburbs?
# Or are these due to the same property being sold multiple times in the recorded timeframe?
#Inspect the duplicate rows and resolve that no rows are entirely duplicate entries.
duplicates = housing_df[housing_df.duplicated(keep=False)]
duplicates


Empty DataFrame
Columns: [Suburb, Address, Rooms, Type, Price, Method, SellerG, Date, Distance, Postcode, Bedroom2, Bathroom, Car, Landsize, BuildingArea, YearBuilt, CouncilArea, Lattitude, Longtitude, Regionname, Propertycount]
Index: []

[0 rows x 21 columns]

In [59]:
#We can see that there are no fully duplicated rows.
# Observe some of the duplicate address values to resolve the causative factor.
duplicates_add = housing_df[housing_df.duplicated(["Address"], keep=False)]
duplicates_add

#We can see that some of the properties, such as 50 Bedford St in Airport West, were sold twice within the experimental period.
# This means the duplicate entries for the same street address are valid. 

Suburb            Address  Rooms Type    Price Method  \
2           Abbotsford       5 Charles St      3    h  1465000     SP   
22          Abbotsford      16 William St      2    h  1310000      S   
53        Airport West      50 Bedford St      3    h   730000     VB   
54        Airport West      50 Bedford St      3    h   770000     SP   
156        Albert Park        6 Brooke St      2    h   980000      S   
...                ...                ...    ...  ...      ...    ...   
18170        Maidstone    88a Ballarat Rd      3    t   706000      S   
18210     Moonee Ponds     55 Pattison St      2    h   700000     VB   
18256  Patterson Lakes      6 Balarang Ct      3    h   670000      S   
18282         Richmond        14 James St      4    h  2700000      S   
18283         Richmond  10/9 Westbank Tce      2    u   410000     PI   

             SellerG       Date  Distance Postcode  ...  Bathroom  Car  \
2             Biggin 2017-04-03       2.5   3067.0  ...       2.0  0.0   
22            Jellis 2016-10-15       2.5   3067.0  ...       1.0  2.0   
53            Nelson 2016-03-12      13.5   3042.0  ...       2.0  1.0   
54            Nelson 2017-04-03      13.5   3042.0  ...       2.0  1.0   
156             Greg 2016-05-28       3.3   3206.0  ...       1.0  0.0   
...              ...        ...       ...      ...  ...       ...  ...   
18170        Sweeney 2017-08-26       6.4   3012.0  ...       NaN  NaN   
18210         Nelson 2017-08-26       6.2   3039.0  ...       NaN  NaN   
18256  hockingstuart 2017-08-26      31.2   3197.0  ...       1.0  3.0   
18282         Jellis 2017-08-26       2.4   3121.0  ...       2.0  1.0   
18283         Jellis 2017-08-26       2.4   3121.0  ...       NaN  NaN   

       Landsize  BuildingArea  YearBuilt    CouncilArea Lattitude  Longtitude  \
2         134.0         150.0     1900.0          Yarra -37.80930   144.99440   
22        238.0          97.0     1890.0          Yarra -37.80900   144.99760   
53          0.0         145.0     1965.0  Moonee Valley -37.72030   144.87550   
54          0.0         145.0     1965.0  Moonee Valley -37.72030   144.87550   
156        79.0          68.0     1890.0   Port Phillip -37.83870   144.95190   
...         ...           ...        ...            ...       ...         ...   
18170       NaN           NaN        NaN            NaN       NaN         NaN   
18210       NaN           NaN        NaN            NaN       NaN         NaN   
18256       NaN         104.0        NaN            NaN -38.07868   145.14743   
18282       NaN         252.0     1900.0            NaN -37.82768   145.00008   
18283       NaN           NaN        NaN            NaN       NaN         NaN   

                       Regionname Propertycount  
2           Northern Metropolitan        4019.0  
22          Northern Metropolitan        4019.0  
53           Western Metropolitan        3464.0  
54           Western Metropolitan        3464.0  
156         Southern Metropolitan        3280.0  
...                           ...           ...  
18170        Western Metropolitan        3873.0  
18210        Western Metropolitan        6232.0  
18256  South-Eastern Metropolitan        3351.0  
18282       Northern Metropolitan       14949.0  
18283       Northern Metropolitan       14949.0  

[512 rows x 21 columns]

This dataset contains 18396 entries, with each entry pertaiing to 1 property sold in Melbourne. There are 21 variables included in this dataset:

1. 'Suburb' - The Melbourne suburb in which the property resides; 330 unique suburbs within this dataset, of the total 1025 suburbs in greater Melbourne. This sample size is due to the dataset being Melbourne City focused.
2. 'Address' - The street address of the property; 18134 unique entires of the 18396 rows, insinuating some duplication of street names across different suburbs, and multiple sales for the same property within the experimental time frame.
3. 'Rooms' - The total number of rooms in the property, excluding bathrooms and other non-living spaces; the dataset hosts an average of 2.96 rooms per property, with a minimum of 1 and a maximum of 12.
4. 'Type' - The type of property; 3 options available:
    - H = House
    - T = Townhouse
    - U = Unit/Apartment
5. 'Price' - The sale price of the property, in AUD (Australian dollars); the dataset hosts an average sale price of AUD$1,056,697, with a minimum of AUD$85,000 and a maximum of AUD$9,000,000.
6. 'Method' - The method of sale used to sell the property; 6 options available: 
    - S = Sold
    - SP = Sold Prior
    - PI = Property Passed In
    - VB = Vendor Bid
    - SA = Sold after Auction
7. 'SellerG' - The real estate agency or individual agent who handled the property sale; 305 unique sellers within this dataset.
8. 'Date' - The date on which the property was sold (yyyy/mm/dd). The dataset hosts 57 individual sale dates, with the earliest property sold on 2016-01-28 and the latest property sold on 2017-12-08.
9. 'Distance' - The distance of the property from the Melbourne Central Business (CBD) in kilometres; the dataset hosts an average distance of 10.39km per property, with a minimum distance of 0km (properties within the CBD) and a maximum distance of 48.10km.
10. 'Postcode' - The Australian postal code allocated to the suburb in which the property resides; 205 unique postcodes within this dataset, of the total ###3532 postcodes in Victoria. This sample size is due to the dataset being Melbourne City focused.
11. 'Bedroom2' - The number of bedrooms in the property, including any non-living spaces that could be used as bedrooms; the dataset hosts an average of 2.91 bedrooms per property, with a minimum of 0 (for studio apartments) and a maximum of 20.
12. 'Bathroom' - The number of bathrooms in the property; the dataset hosts an average of 1.54 bathrooms per property, with a minimum of 0 (for studio apartments) and a maximum of 8.
13. 'Car' - The number of car parking spaces associated with the property; the dataset hosts an average of 1.62 car spaces per property, with a minimum of 0 and a maximum of 10.
14. 'Landsize' - The size of the property's land, in square meters (m^2); the dataset hosts an average land size of 558.12m^2 per property, with a minimum landsize of 0m^2 and a maximum landsize of 433,014m^2.
15. 'BuildingArea' - The total area of the property's building, in square meters (m^2); the dataset hosts an average buidling area of 151.22m^2 per property, with a minimum building area of 0m^2 (land-only sales) and a maximum building area of 44,515m^2.
16. 'YearBuilt' - The year in which the property building was built; the dataset hosts an average year built of 1965.87, with the earliest property built in 1196 and the latest propert built in 2018.
17. 'CouncilArea' - The name of local government area in which the property resides (contains postcodes); 33 unique council areas within this dataset, of the total 79 council areas in Victoria. This sample size is due to the dataset being Melbourne City focused.
18. 'Latitude' - The degrees of distance north or south of the equator at which the property is located; the dataset hosts an average Latitude of -37.81 with a minimum of -38.18 and a maximum of 37.40.
19. 'Longtitude' - The degrees of distance east or west of the prime meridian at which the property is located; the dataset hosts an average Longtitude of 145.00 with a minimum of 144.43 and a maximum of 145.53.
20. 'Regionname' - The name of the region within which the property is located (contains LGAs); 6 options:
    - Southern Metropolitan
    - Northern Metropolitan
    - Western Metropolitan
    - Eastern Metropolitan
    - South-Eastern Metropolitan
    - Eastern Victoria
    - Northern Victoria
    - Western Victoria
21. 'Propertycount' - The number of properties within the suburb that the property resides in; the dataset hosts an average of 7517.98 properties per suburb, with a minimum of 249 and a maximum of 21650.

In [60]:
#Export the cleaned dataset as a csv file.
housing_df.to_csv("Properties of Melbourne City (CLEANED).csv")
